morphct/kmc_analyze.py

```diff
@@ -0,0 +1,2165 @@
+    ):
+        hop_name = f"{hop_type}_{target}r{sp}"
+        n_hops = len(prop_lists[hop_name])
+        if n_hops == 0:
+            continue
+
+        other_hop = hop_types[hop_types.index(hop_type) * -1 + 1]
+        other_hop_name = f"{other_hop}_{target}r{sp}"
+
+        total_hops = n_hops + len(prop_lists[other_hop_name])
+        proportion = n_hops / total_hops
+
+        mean_rate = np.mean(prop_lists[hop_name])
+        stdev_rate = np.std(prop_lists[hop_name])
+        data_dict[f"{hop_name}_hops"] = n_hops
+        data_dict[f"{hop_name}_proportion"] = proportion
+        data_dict[f"{hop_name}_rate_mean"] = mean_rate
+        data_dict[f"{hop_name}_rate_std"] = stdev_rate
+
+
+def plot_stacked_hist_rates(
+    data1, data2, labels, species, path
+):  # pragma: no cover
+    """
+
+    Parameters
+    ----------
+    path : path
+        Path to directory where to save the plot.
+
+    Returns
+    -------
+    """
+    plt.figure()
+    n, bins, patches = plt.hist(
+        [data1, data2],
+        bins=np.logspace(1, 18, 40),
+        stacked=True,
+        color=["r", "b"],
+        label=labels,
+    )
+    plt.ylabel("Frequency (Arb. U.)")
+    plt.xlabel(rf"{species.capitalize()} k$_{{i,j}}$ (s$^{-1}$)")
+    plt.xlim([1, 1e18])
+    plt.xticks([1e0, 1e3, 1e6, 1e9, 1e12, 1e15, 1e18])
+    plt.ylim([0, np.max(n) * 1.02])
+    plt.legend(loc=2, prop={"size": 18})
+    plt.gca().set_xscale("log")
+    filename = f"{species}_hopping_rate_clusters.png"
+    filepath = os.path.join(path, filename)
+    plt.savefig(filepath, dpi=300)
+    plt.close()
+    print(f"\tFigure saved as {filename}")
+
+
+def plot_stacked_hist_tis(
+    data1, data2, labels, species, cutoff, path
+):  # pragma: no cover
+    """
+
+    Parameters
+    ----------
+    path : path
+        Path to directory where to save the plot.
+
+    Returns
+    -------
+    """
+    plt.figure()
+    n, bins, patches = plt.hist(
+        [data1, data2],
+        bins=np.linspace(0, 1.2, 20),
+        stacked=True,
+        color=["r", "b"],
+        label=labels,
+    )
+    plt.ylabel("Frequency (Arb. U.)")
+    plt.xlabel(rf"{species.capitalize()} J$_{{i,j}}$ (eV)")
+    plt.ylim([0, np.max(n) * 1.02])
+    if cutoff is not None:
+        plt.axvline(float(cutoff), c="k")
+    plt.legend(loc=0, prop={"size": 18})
+
+    filename = f"{species}_transfer_integral_clusters.png"
+    filepath = os.path.join(path, filename)
+    plt.savefig(filepath, dpi=300)
+    plt.close()
+    print(f"\tFigure saved as {filename}")
+
+
+def write_csv(data_dict, path): # pragma: no cover
+    """
+
+    Parameters
+    ----------
+    path : path
+        Path to directory where to save the plot.
+
+    Returns
+    -------
+    """
+    filepath = os.path.join(path, "results.csv")
+    with open(filepath, "w") as f:
+        w = csv.writer(f)
+        for key, val in data_dict.items():
+            w.writerow([key, val])
+    print(f"\tCSV file written to {filepath}")
+
+
+# TODO EJ
```

Comment:
I don't have a quick answer for this one. It looks like it's for finding distances to cut off distributions if a cutoff doesn't exist for one already, but this'll take some spaghetti-pulling to answer fully.

Reply:
OK, I have made an issue to remind me to look into this more.